Userland/Libraries/LibWeb/DOM/Node.cpp

```diff
@@ -2312,11 +2314,18 @@ ErrorOr<String> Node::name_or_description(NameOrDescription target, Document con
             // - Otherwise, return the value of aria-label.
             return element->aria_label().value();
         }
-        // TODO: D. Otherwise, if the current node's native markup provides an attribute (e.g. title) or element (e.g. HTML label) that defines a text alternative,
-        //      return that alternative in the form of a flat string as defined by the host language, unless the element is marked as presentational (role="presentation" or role="none").
+        // E. Host Language Label: Otherwise, if the current node's native markup provides
+        // an attribute (e.g. alt) or element (e.g. HTML label or SVG title) that defines
+        // a text alternative, return that alternative in the form of a flat string as
+        // defined by the host language, unless the element is marked as presentational
+        // (role="presentation" or role="none").
+        if (is<HTML::HTMLImageElement>(*element))
+            if (auto const& img = static_cast<HTML::HTMLImageElement const&>(*element);
+                img.has_attribute(HTML::AttributeNames::alt)
+                && role != ARIA::Role::presentation && role != ARIA::Role::none)
```

Comment:
This is a bit chonky for a declaration-and-condition `if` statement IMO, I would break it up to make it easier to read, do the cheaper `role` checks earlier, and we don't need to look up the `alt` attribute twice.

We should also add a FIXME about the other element types noted in the spec step.

Something like:

```c++
if (role != ARIA::Role::presentation && role != ARIA::Role::none) {
    if (is<HTML::HTMLImageElement>(*element)) {
        if (auto alt = element->get_attribute(HTML::AttributeNames::alt); alt.has_value())
            return alt.release_value();
    }

    // FIXME: Handle other element types like HTMLLabelElement and SVGTitleElement.
}
```

Reply:
> This is a bit chonky for a declaration-and-condition `if` statement IMO, I would break it up to make it easier to read, do the cheaper `role` checks earlier, and we don't need to look up the `alt` attribute twice.

Thanks much — definitely a better, cleaner, clearer way to handle it

> We should also add a FIXME about the other element types noted in the spec step.
> …
> ```c++
>     // FIXME: Handle other element types like HTMLLabelElement and SVGTitleElement.
> }
> ```

See the (more-verbose) comment I added; it’s not clear to me yet that the existing code isn’t _already_ the `HTMLLabelElement` case in conformance with the spec requirements (and the WPT tests).

More specifically: The existing code code handles the `label` element case by handling any of the `label` element’s descendant nodes that need to have an accessible name exposed. So in those cases, the accessible name from the `label` content is exposed by those descendant nodes.

It’s not clear to me yet whether the spec requires an accessible name to _also_ be separately exposed on the `label` element _itself_. Gecko exposes one for the `label` itself, but WebKit definitely doesn’t — and it looks to me like Blink doesn’t either.

So… I’m not sure which other UAs are right and which are wrong. But I plan to ask — in some W3C accessibility fora — both some knowledgeable accessibility practitioners (to find out what the ideal behavior is from the POV of actual screen-reader/AT users) and also the actual spec editors (to find out what the intent of the spec is).